### ISSUE TITLE
Panel flickers since 1.9.0.0

### ISSUE THREAD

**Issue by ginga81** [leezer3/OpenBVE#882]
### Description

As the title suggests, the Panel display becomes flickers.
He uses the huge Needle as he says that the size is 6892.
Run OpenBVE
Choose any route.
The train that is his Series-E2,E5,E6 or any Shinkansen.
When the train driving, center ATC panel becomes flickering.

This time, I was able to know in advance that the next release would be out, so I asked him and decided that it would be better to have him incorporate the fix before the official release.

### Reproduction

The Route is any.
The train that is his Series-E2,E5,E6 or any Shinkansen.

## Route

The Route is any.

## Train

For example:
Series-200
https://www.youtube.com/watch?v=uQSpO23-Z6s
FASTECH360Z
https://youtu.be/Ky9_6A8sXvk
Series-E2-0
https://youtu.be/GMZaGCPvymk
Or any his Shinkansen's trains.

## Logs

This issue isn't crash. So no log.

### Related information 

My PC's spec:
OS:Ubuntu 22.04 LTS 64bit
CPU:AMD® Ryzen 7 2700x eight-core processor × 16
Memory:31.3GB
Graphic:AMD Radeon RX 590 Series (polaris10, LLVM 15.0.6, DRM 3.47, 5.19.0-35-generic) (0x67df)

**Comment by leezer3** [comment]
Probably already fixed, i think it'll be this:
https://github.com/leezer3/OpenBVE/issues/879

**Comment by ginga81** [comment]
I have tested and reproduced on 1.9.0.2.
I don't know if the logic is the same, but it's a Panel and not an Object.
I have not confirmed any object transparency anomalies.
Even the daily build of 2023.03.29 still flickers, so I don't think it's fixed.
https://user-images.githubusercontent.com/22734699/228910369-a7989520-7130-476d-91f6-c6c9f2daffdc.mp4

**Comment by leezer3** [comment]
Not seeing any flickering here I'm afraid.
https://user-images.githubusercontent.com/7688245/229083632-3ce6f849-6b8c-43d1-ba81-266503e8a120.mp4

I've tried with all transparency settings, GL1.2 and GL3 on Windows at the minute.

**Comment by leezer3** [comment]
Getting somewhere-
TEC_G6 doesn't flicker, but TEC_E2_0_gto does.

The flickering is being caused by the DigitalGauge.
These seem to first break here (albeit differently):
https://github.com/leezer3/OpenBVE/commit/577500f4273cfd97288667794eb9200ca1bc79b0

**Comment by leezer3** [comment]
Interesting...

That commit switched to face ordering via a unified method.
This is the original sort method for overlay faces:
https://github.com/leezer3/OpenBVE/blob/577500f4273cfd97288667794eb9200ca1bc79b0/source/LibRender2/Objects/ObjectLibrary.cs#L297

Looking at the code, this sort method is broken, and actually doing nothing.

Still investigating at the minute.

**Comment by leezer3** [comment]
Really odd....

The sort algorithm there is definitely dodgy, but moving the face sort for the cab to immediately before drawing seems to solve it here.
I can't immediately explain this, but see if the build from today is any better.

**Comment by ginga81** [comment]
I tried with the 20230331 build, but neither E6 nor E2-0 were fixed.
It may not be reflected yet, so I wait a few hours and try again.

**Comment by leezer3** [comment]
It shouldn't be.
The commits with the green tick are those which it's built against:
https://github.com/leezer3/OpenBVE/commits/master

Last built commit was the 29th, so the one that's uploaded is definitely right.
Might be something else going on (presumably thread timing related somehow), will have to try and see if I can find anything else- I don't like the change I made at all, it shouldn't have done anything.....

**Comment by leezer3** [comment]
Yeah, it's the commit I linked above that is part of the cause.

The trouble is that this seems to have been masking a different issue, which is by the looks of things timing related.

**Comment by leezer3** [comment]
http://vps.bvecornwall.co.uk/OpenBVE/Experimental/flicker.zip

Please see if this flickers.
I've discovered [this](https://github.com/leezer3/OpenBVE/commit/136d599ac97c205feababce48b3b81c23ec28849) from a couple of years ago, and I'm wondering if something was broken then, and it shouldn't have been added.

**Comment by leezer3** [comment]
https://vps.bvecornwall.co.uk/OpenBVE/Builds/OpenBVE-2023-04-02.zip

Another approach entirely to the flickering.
This is possibly something to do with the interrelationship between the QuickSort and Stable sort algorithms, but isn't reliable at all....

TLDR:
The original approach, which I've just restored modified the original copy of the list when sorting. Wondering if this has an effect when new items are inserted into the list....

**Comment by ginga81** [comment]
I tested with this build at E2-0 E5 E6 E7 200 E955. 
It was displayed without any problems!
I think this is no problem, what do you think?

**Comment by leezer3** [comment]
Sounds good.

I'll try and push a new public build in the next few days, as I think getting to the bottom of that is quite important.
